File: FirmwareGO/webapp/controllers/ApplicationStateController.go
package controllers

import (
	. "FirmwareGO/application/services"
	"FirmwareGO/webapp/middleware/prometheus"
	"FirmwareGO/webapp/status"
	"github.com/gin-gonic/gin"
	"net/http"
)

type ApplicationStateController struct {
	ApplicationStateService *ApplicationStateService
}

func NewApplicationStateController(applicationStateService *ApplicationStateService) *ApplicationStateController {
	return &ApplicationStateController{ApplicationStateService: applicationStateService}
}

func (controller *ApplicationStateController) RegisterRoutes(router *gin.Engine) {
	router.GET("/health", controller.HealthCheck)
	router.GET("/ready", controller.ReadyCheck)
}

func (controller *ApplicationStateController) HealthCheck(context *gin.Context) {
	prometheus.SetHealthStatus(true)
	context.JSON(http.StatusOK, gin.H{"status": "up"})
}

func (controller *ApplicationStateController) ReadyCheck(context *gin.Context) {
	if controller.ApplicationStateService.DbIsConnected() {
		if status.IsMigrationSuccessful() {
			prometheus.SetReadinessStatus(true)
			context.JSON(http.StatusOK, gin.H{"status": "ready"})
		} else {
			prometheus.SetReadinessStatus(false)
			context.JSON(http.StatusServiceUnavailable, gin.H{"status": "failed to apply pending database migrations."})
		}
	} else {
		prometheus.SetReadinessStatus(false)
		context.JSON(http.StatusServiceUnavailable, gin.H{"status": "could not connect to database."})
	}
}

File: NotificationMS/WebApp/Controllers/ApplicationStateController.cs
using Application.ApplicationServices.Interfaces;
using Microsoft.AspNetCore.Mvc;
using WebApp.Middleware.Prometheus;
using WebApp.Middleware.Status;

namespace WebApp.Controllers;

[ApiController]
public class ApplicationStateController : ControllerBase
{
    private readonly IApplicationStateService _applicationStateService;
    private readonly MigrationStatus _migrationStatus;
    private readonly CustomMetrics _customMetrics;

    public ApplicationStateController(IApplicationStateService applicationStateService, MigrationStatus migrationStatus,
        CustomMetrics customMetrics)
    {
        _applicationStateService = applicationStateService;
        _migrationStatus = migrationStatus;
        _customMetrics = customMetrics;
    }

    [HttpGet("/health")]
    [ApiExplorerSettings(IgnoreApi = true)]
    public async Task<ActionResult<bool>> HealthCheck()
    {
        // Checks whether application is up and running so OK is immediately returned
        _customMetrics.HealthCheckGauge.Set(1);
        return Ok();
    }

    [HttpGet("/ready")]
    [ApiExplorerSettings(IgnoreApi = true)]
    public async Task<IActionResult> ReadyCheck()
    {
        if (await _applicationStateService.DbIsConnected())
        {
            if (_migrationStatus.IsMigrationSuccessful)
            {
                _customMetrics.ReadinessCheckGauge.Set(1);
                return Ok("status: ready");
            }

            _customMetrics.ReadinessCheckGauge.Set(0);
            return StatusCode(503, "status: failed to apply pending database migrations.");
        }

        _customMetrics.ReadinessCheckGauge.Set(0);
        return StatusCode(503, "status: could not connect to database");
    }
}